src/workerd/jsg/resource.h

```diff
@@ -998,19 +1372,41 @@ struct ResourceTypeBuilder {
   template <const char* name, typename Getter, Getter getter, typename Setter, Setter setter>
   inline void registerPrototypeProperty() {
     auto v8Name = v8StrIntern(isolate, name);
+    v8::Local<v8::FunctionTemplate> getterFn;
+    v8::Local<v8::FunctionTemplate> setterFn;
 
     using Gcb = PropertyGetterCallback<TypeWrapper, name, Getter, getter, isContext>;
+    using Scb = PropertySetterCallback<TypeWrapper, name, Setter, setter, isContext>;
     if (!Gcb::enumerable) {
       inspectProperties->Set(v8Name, v8::False(isolate), v8::PropertyAttribute::ReadOnly);
     }
 
-    // Note that we cannot use `SetNativeDataProperty(...)` here the way we do with other
-    // properties because it will not properly handle the prototype chain when it comes
-    // to using setters... which is annoying. It means we end up having to use FunctionTemplates
-    // instead of the more convenient property callbacks.
-    auto getterFn = v8::FunctionTemplate::New(isolate, Gcb::callback);
-    auto setterFn = v8::FunctionTemplate::New(
-        isolate, &PropertySetterCallback<TypeWrapper, name, Setter, setter, isContext>::callback);
+    bool useSlowApi = true;
+    if constexpr (isFastMethodCompatible<TypeWrapper, Getter> &&
+        isFastMethodCompatible<TypeWrapper, Setter>) {
+      if (util::Autogate::isEnabled(util::AutogateKey::V8_FAST_API)) {
```

Comment:
Maybe it's worth initializing a boolean in TypeWrapper based on the autogate and checking that, rather than calling `isEnabled()` everywhere. Normally `isEnabled()` is cheap enough, but _every_ JSG_METHOD is going to be generating these calls and they are all out-of-line calls, it could actually be non-negligible code bloat.

Also it would be kind of nice not to introduce the dependency on autogates from JSG, we could instead pass it down as a parameter from a higher level.

Reply:
My latest commit (https://github.com/cloudflare/workerd/pull/4008/commits/a4336a0e81a35b4e9c29de927b49df3ac36e98f2) should address this. Thank you for the review.